src/Engine/Data/EnvironmentTexture.cpp

```diff
@@ -0,0 +1,739 @@
+            auto z          = std::cos( theta );
+            float* thePixel = getPixel( x, y, z );
+            updateCoeffs( thePixel, -x, y, z, std::sin( theta ) * dtheta * dphi );
+        }
+    }
+    tomatrix();
+}
+
+void EnvironmentTexture::updateCoeffs( float* hdr, float x, float y, float z, float domega ) {
+    /******************************************************************
+       Update the coefficients (i.e. compute the next term in the
+       integral) based on the lighting value hdr[3], the differential
+       solid angle domega and cartesian components of surface normal x,y,z
+
+       Inputs:  hdr = L(x,y,z) [note that x^2+y^2+z^2 = 1]
+                i.e. the illumination at position (x,y,z)
+
+                domega = The solid angle at the pixel corresponding to
+                (x,y,z).  For these light probes, this is given by
+
+                x,y,z  = Cartesian components of surface normal
+
+       Notes:   Of course, there are better numerical methods to do
+                integration, but this naive approach is sufficient for our
+                purpose.
+    *********************************************************************/
+    for ( int col = 0; col < 3; col++ ) {
+        float c; /* A different constant for each coefficient */
+
+        /* L_{00}.  Note that Y_{00} = 0.282095 */
+        c = 0.282095f;
+        m_shcoefs[0][col] += hdr[col] * c * domega;
+
+        /* L_{1m}. -1 <= m <= 1.  The linear terms */
+        c = 0.488603f;
+        m_shcoefs[1][col] += hdr[col] * ( c * y ) * domega; /* Y_{1-1} = 0.488603 y  */
+        m_shcoefs[2][col] += hdr[col] * ( c * z ) * domega; /* Y_{10}  = 0.488603 z  */
+        m_shcoefs[3][col] += hdr[col] * ( c * x ) * domega; /* Y_{11}  = 0.488603 x  */
+
+        /* The Quadratic terms, L_{2m} -2 <= m <= 2 */
+
+        /* First, L_{2-2}, L_{2-1}, L_{21} corresponding to xy,yz,xz */
+        c = 1.092548f;
+        m_shcoefs[4][col] += hdr[col] * ( c * x * y ) * domega; /* Y_{2-2} = 1.092548 xy */
+        m_shcoefs[5][col] += hdr[col] * ( c * y * z ) * domega; /* Y_{2-1} = 1.092548 yz */
+        m_shcoefs[7][col] += hdr[col] * ( c * x * z ) * domega; /* Y_{21}  = 1.092548 xz */
+
+        /* L_{20}.  Note that Y_{20} = 0.315392 (3z^2 - 1) */
+        c = 0.315392f;
+        m_shcoefs[6][col] += hdr[col] * ( c * ( 3 * z * z - 1 ) ) * domega;
+
+        /* L_{22}.  Note that Y_{22} = 0.546274 (x^2 - y^2) */
+        c = 0.546274f;
+        m_shcoefs[8][col] += hdr[col] * ( c * ( x * x - y * y ) ) * domega;
+    }
+}
+
+void EnvironmentTexture::tomatrix( void ) {
+    /* Form the quadratic form matrix (see equations 11 and 12 in paper) */
+    int col;
+    float c1, c2, c3, c4, c5;
+    c1 = 0.429043;
+    c2 = 0.511664;
+    c3 = 0.743125;
+    c4 = 0.886227;
+    c5 = 0.247708;
+
+    for ( col = 0; col < 3; col++ ) { /* Equation 12 */
+
+        m_shMatrices[col]( 0, 0 ) = c1 * m_shcoefs[8][col]; /* c1 L_{22}  */
```

Comment:
The following loop could take benefit of the comma initializer (https://eigen.tuxfamily.org/dox/group__TutorialAdvancedInitialization.html), to avoid calling `()` access operator multiple times, this should save time.

Reply:
I'm not a great fan of this comma initializer that requires some formatting to prevent warning from static analysers but, well, done.